packages/vite/src/node/build.ts

```diff
@@ -466,34 +466,26 @@ export async function resolveBuildPlugins(config: ResolvedConfig): Promise<{
   pre: Plugin[]
   post: Plugin[]
 }> {
-  const { commonjsOptions } = config.build
-  const usePluginCommonjs =
-    !Array.isArray(commonjsOptions.include) ||
-    commonjsOptions.include.length !== 0
   return {
     pre: [
       completeSystemWrapPlugin(),
-      /**
-       * environment.config.build.commonjsOptions isn't currently supported
-       * when builder.sharedConfigBuild or builder.sharedPlugins enabled.
-       * To do it, we could inject one commonjs plugin per environment with
-       * an applyToEnvironment hook.
-       */
-      ...(usePluginCommonjs ? [commonjsPlugin(commonjsOptions)] : []),
+      perEnvironmentPlugin('commonjs', (environment) => {
+        const { commonjsOptions } = environment.config.build
+        const usePluginCommonjs =
+          !Array.isArray(commonjsOptions.include) ||
+          commonjsOptions.include.length !== 0
+        return usePluginCommonjs ? commonjsPlugin(commonjsOptions) : false
+      }),
       dataURIPlugin(),
-      /**
-       * environment.config.build.rollupOptions.plugins isn't supported
-       * when builder.sharedConfigBuild or builder.sharedPlugins is enabled.
-       * To do it, we should add all these plugins to the global pipeline, each with
-       * an applyToEnvironment hook. It is similar to letting the user add per
-       * environment plugins giving them a environment.config.plugins option that
-       * we decided against.
-       * For backward compatibility, we are still injecting the rollup plugins
-       * defined in the default root build options.
-       */
-      ...((
-        await asyncFlatten(arraify(config.build.rollupOptions.plugins))
-      ).filter(Boolean) as Plugin[]),
+      perEnvironmentPlugin(
+        'vite:rollup-options-plugins',
```

Comment:
Neat trick 👍

Reply:
Actually, wouldn't the function still return the same instance of the Rollup plugins and are being shared across environments, which results in the same behaviour?

Reply:
Yes, if the plugin is set at the root `build.rollupOptions.plugins` then it is the same as before. But now it properly works if someone uses `env.build.rollupOptions.plugins`. There are some rollup plugins that is ok to have in the root shared, so maybe it is fine. The other option would be define that the root only applies to the client but that is a breaking change. I think that could be the best in this case, probably there isn't much usage of `build.rollupOptions.plugins` out there?

Reply:
Actually, I think the plugins at root being shared is the most intuitive behavior aligning with the `plugins` top level option. The user should add plugins that can't be shared at top level (note: all this is only for users that set `builder.sharedConfig`)

Reply:
Yeah I was thinking more of the top-level `build.rollupOptions.plugins` which would still work like before. If the user puts plugins under `env.build.rollupOptions.plugins`, wouldn't the instance be already tied to that environment's config plugins, so we can directly spread it out like before? Or maybe I got confused here.

Reply:
We can't do this with the current way in which plugins are resolved. The `rollupOptions.plugins` are injected in the middle of the normal build plugins [here](https://github.com/vitejs/vite/blob/fb227ec4402246b5a13e274c881d9de6dd8082dd/packages/vite/src/node/build.ts#L494). So if we want to let people keep messing with the plugins in `resolveConfig` (they shouldn't but some do), we need some kind of marker. A plugin with the `applyToEnvironment()` hook does this for us now.

Reply:
I think I understand it better now 👍